lib/logstash/inputs/twitter.rb

```diff
@@ -74,51 +104,84 @@ def on_body(data)
       end
     end
 
-    @client = Twitter::Streaming::Client.new do |c|
-      c.consumer_key = @consumer_key
-      c.consumer_secret = @consumer_secret.value
-      c.access_token = @oauth_token
-      c.access_token_secret = @oauth_token_secret.value
-    end
+    @rest_client     = Twitter::REST::Client.new       { |c|  configure(c) }
+    @stream_client   = Twitter::Streaming::Client.new  { |c|  configure(c) }
+    @twitter_options = build_options
   end
 
-  public
   def run(queue)
-    @logger.info("Starting twitter tracking", :keywords => @keywords)
+    @logger.info("Starting twitter tracking", twitter_options)
     begin
-      @client.filter(:track => @keywords.join(",")) do |tweet|
-        return if stop?
-        if tweet.is_a?(Twitter::Tweet)
-          event = from_tweet(tweet)
-          decorate(event)
-          queue << event
+      if @use_samples
+        @stream_client.sample do |tweet|
+          return if stop?
+          tweet_processor(queue, tweet)
         end
-      end # client.filter
+      else
+        @stream_client.filter(twitter_options) do |tweet|
+          return if stop?
+          tweet_processor(queue, tweet)
+        end
+      end
     rescue Twitter::Error::TooManyRequests => e
       @logger.warn("Twitter too many requests error, sleeping for #{e.rate_limit.reset_in}s")
       Stud.stoppable_sleep(e.rate_limit.reset_in) { stop? }
       retry
     rescue => e
-      @logger.warn("Twitter client error", :message => e.message, :exception => e, :backtrace => e.backtrace)
+      @logger.warn("Twitter client error", :message => e.message, :exception => e, :backtrace => e.backtrace, :options => @filter_options)
       retry
     end
   end # def run
 
+  def stop
+    @stream_client = nil
+  end
+
+  def twitter_options
+    @twitter_options.delete(:level) # This is added by the logger when you pass the full hash as params
```

Comment:
Hmmmm, maybe we should default to cloning it in that case? 

This is the part where I wish we were using immutable datastructures in clojure :(